lib/self-evict.js

```diff
@@ -0,0 +1,303 @@
+    /**
+     * The history of phases
+     * @type {[Phase]}
+     */
+    this.phases = [];
+}
+
+/**
+ * Get the current  phase or null when self eviction isn't initiated
+ * @return {?Phase}
+ */
+SelfEvict.prototype.currentPhase = function currentPhase() {
+    if (this.phases && this.phases.length > 0) {
+        return this.phases[this.phases.length - 1];
+    }
+    return null;
+};
+
+/**
+ * Register hooks in the eviction sequence. At least one of hook-methods should
+ * be defined.
+ *
+ * @param {Object} hooks The hooks to register
+ * @param {string} hooks.name The name of the registered hooks
+ * @param {SelfEvict~Hook} [hooks.preEvict] The preEvict hook to register.
+ * @param {SelfEvict~Hook} [hooks.postEvict] The postEvict hook to register.
+ */
+SelfEvict.prototype.registerHooks = function registerHooks(hooks) {
+    if (!hooks) {
+        throw errors.ArgumentRequiredError({argument: 'hooks'});
+    }
+
+    if (!hooks.name) {
+        throw errors.FieldRequiredError({argument: 'hooks', field: 'name'});
+    }
+    var name = hooks.name;
+
+    if (this.preEvictHooks[name] || this.postEvictHooks[name]) {
+        throw errors.DuplicateHookError({name: name});
+    }
+
+    if (!hooks.preEvict && !hooks.postEvict) {
+        throw errors.MethodRequiredError({
+            argument: 'hooks',
+            method: 'preEvict and/or postEvict'
+        });
+    }
+
+    if (hooks.preEvict) {
+        this.preEvictHooks[name] = hooks.preEvict;
+    }
+
+    if (hooks.postEvict) {
+        this.postEvictHooks[name] = hooks.postEvict;
+    }
+};
+
+/**
+ * This callback is called after self eviction is complete. When the callback is
+ * invoked, it's safe to shut down the ringpop service or destroy ringpop.
+ *
+ * @callback SelfEvict~callback
+ * @param {Error} [err] The error if one occurred
+ * @see Ringpop#destroy
+ */
+
+/**
+ * Initiate the self eviction sequence.
+ * @param {SelfEvict~callback} callback called when self eviction is complete.
+ */
+SelfEvict.prototype.initiate = function initiate(callback) {
+    if (this.currentPhase() !== null) {
+        var error = RingpopIsAlreadyShuttingDownError({phase: this.currentPhase()});
+        this.ringpop.logger.warn('ringpop is already shutting down', {
+            local: this.ringpop.whoami(),
+            error: error
+        });
+        callback(error);
+        return;
+    }
+    this.ringpop.logger.info('ringpop is initiating self eviction sequence', {
+        local: this.ringpop.whoami()
+    });
+    this.callback = callback;
+
+    this._preEvict();
+};
+
+SelfEvict.prototype._preEvict = function _preEvict() {
+    this._transitionTo(PhaseNames.PreEvict);
+
+    var self = this;
+    this._runHooks(this.preEvictHooks, function hooksDone() {
+        process.nextTick(self._evict.bind(self));
+    });
+};
+
+SelfEvict.prototype._evict = function _evict() {
+    this._transitionTo(PhaseNames.Evicting);
+
+    this.ringpop.membership.setLocalStatus(Member.Status.faulty);
```

Comment:
I think we should forcefully ping nodes here. If there are no posthooks or they take a really short time our faulty declaration might not reach any node before shutting down. This will cause the normal suspect period to kick in.

As discussed the number and/or targets to ping should be configurable. You might want to ping all the new owners of part of your keyspace. You might also decide to quickly send 3 pings to get the information into the network before disappearing.

Only after you have confirmation that the nodes have received the gossip it is safe to transition to the postEviction state I believe.